### PR TITLE
Connect Worker to Backend

### PR DESCRIPTION
**Pull request by pajowu**

### CODE REVIEW

File: proto/transcribee_proto/api.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+
+class TaskBase(BaseModel):
+    id: str
+    document: Document
+    task_type: TaskType
+
+
+class DiarizeTask(TaskBase):
+    task_type: Literal[TaskType.DIARIZE]
```
Comment (anuejn) [inline proto/transcribee_proto/api.py:30]: there should be a `= TaskType.DIARIZE`

File: proto/transcribee_proto/api.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    task_parameters: Dict[str, Any]
+
+
+class TranscribeTaskParameters(BaseModel):
+    lang: str
+    model: str
+
+
+class TranscribeTask(TaskBase):
+    task_type: Literal[TaskType.TRANSCRIBE]
```
Comment (anuejn) [inline proto/transcribee_proto/api.py:40]: see above

File: proto/transcribee_proto/api.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    task_parameters: TranscribeTaskParameters
+
+
+class AlignTask(TaskBase):
+    task_type: Literal[TaskType.ALIGN]
```
Comment (anuejn) [inline proto/transcribee_proto/api.py:45]: see above

File: worker/run.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import logging
+
+from transcribee_proto.api import TaskType
+from transcribee_worker.worker import Worker
+
+logging.basicConfig(level=logging.INFO)
+
+
+async def main():
+    parser = argparse.ArgumentParser(description="Worker")
```
Comment (anuejn) [inline worker/run.py:14]:
```suggestion
    parser = argparse.ArgumentParser(description="the worker doing ML inference for the transcribee open source transcription platform")
```

File: worker/transcribee_worker/worker.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+import requests
+from pydantic import parse_raw_as
+from transcribee_proto.api import AlignTask, AssignedTask, DiarizeTask
+from transcribee_proto.api import Document as ApiDocument
+from transcribee_proto.api import TaskType, TranscribeTask
+from transcribee_worker.util import load_audio
+from transcribee_worker.whisper_transcribe import transcribe
+
+logging.basicConfig(level=logging.INFO)
```
Comment (anuejn) [inline worker/transcribee_worker/worker.py:16]: this should be in the run.py

File: worker/run.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    )
+    parser.add_argument("--token", help="Worker token", required=True)
+    args = parser.parse_args()
+
+    worker = Worker(
+        base_url=f"{args.coordinator}/api/v1/tasks",
+        token=args.token,
+        task_types=[TaskType.TRANSCRIBE],
+    )
+    await worker.run_task()
```
Comment (anuejn) [inline worker/run.py:28]: can I have a while true?
also can I have sleeping if no task was executed?

**Review by anuejn — CHANGES_REQUESTED** [review]

File: worker/run.py (hunk 4 above)
Comment (pajowu) [reply worker/run.py:14]: I changed it to `"The worker for the transcribee open source transcription platform"`, since we not (necessarily) only do ml inference in the worker

File: worker/transcribee_worker/whisper_transcribe.py [hunk 7]
```diff
@@ -123,39 +124,49 @@ def handle_new_segment(
...
-    params.no_context = (
-        False  # if False, feeds back already transcribed text back to the model
+    params = (
+        api.Params.from_sampling_strategy(sampling)
+        .with_no_context(
+            False
+        )  # if False, feeds back already transcribed text back to the model
+        .with_num_threads(4)
+        .with_language(lang_code)
+        .with_max_segment_length(120)  # parameter stolen from whisper.cpp cli
```
Comment (rroohhh) [inline worker/transcribee_worker/whisper_transcribe.py:144]: Why is this changed from 60 to 120?

File: worker/transcribee_worker/worker.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+            return
+
+        document_file = self.get_document_file(task.document)
+        audio = load_audio(str(document_file))
+
+        def progress_callback(_ctx, progress, _data):
+            self.keepalive(task.id, progress=progress / 100)
+
+        paragraphs = []
+        async for paragraph in transcribe(
```
Comment (rroohhh) [inline worker/transcribee_worker/worker.py:103]: This could use async list comprehension

File: worker/transcribee_worker/worker.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+        )
+        req.raise_for_status()
+        return parse_raw_as(Optional[AssignedTask], req.text)
+
+    def _get_tmpfile(self, filename: str) -> Path:
+        if self.tmpdir is None:
+            raise ValueError("`tmpdir` must be set")
+        return self.tmpdir / filename
+
+    def get_document_file(self, document: ApiDocument) -> Optional[Path]:
```
Comment (rroohhh) [inline worker/transcribee_worker/worker.py:55]: Why do we save the file to the filesystem? 
Can we not just keep it in memory?

File: worker/transcribee_worker/worker.py (hunk 8 above)
Comment (anuejn) [reply worker/transcribee_worker/worker.py:103]: I belive it is fine for now as the worker throws a `NotImplemented` anyways currently.

File: proto/transcribee_proto/document.py [hunk 10]
```diff
@@ -32,6 +32,12 @@ class Document(BaseModel):
     lang: str
     paragraphs: List[Paragraph]
 
+    def is_empty(self) -> bool:
```
[the remark this answers is not in the document]
Comment (anuejn) [reply proto/transcribee_proto/document.py:35]: I would argue it should not contain True in this case. No matter if it is whitespace there is *something* and it is thus not empty?

File: worker/transcribee_worker/whisper_transcribe.py (hunk 7 above)
Comment (anuejn) [reply worker/transcribee_worker/whisper_transcribe.py:144]: Can we just set this to 0? I heard this will give best results?

**Review by anuejn — APPROVED** [review]
I am okay with this in the current state.
If it is okay for you @pajowu and @rroohhh plz merge :)

File: worker/transcribee_worker/whisper_transcribe.py (hunk 7 above)
Comment (rroohhh) [reply worker/transcribee_worker/whisper_transcribe.py:144]: Where did you hear this?
Why does the whisper.cpp cli set it to `60` when one specifies `0` then? (ref https://github.com/ggerganov/whisper.cpp/blob/09e90680072d8ecdf02eaf21c393218385d2c616/examples/main/main.cpp#L590)

File: worker/run.py [hunk 11]
```diff
@@ -28,8 +30,17 @@ async def main():
...
-        no_work = await worker.run_task()
-        if no_work:
+        try:
+            no_work = await worker.run_task()
+            if no_work:
+                await asyncio.sleep(5)
+        except requests.exceptions.ConnectionError:
+            logging.warn("could not connect to backend")
+            await asyncio.sleep(5)
+        except Exception:
```
Comment (rroohhh) [inline worker/run.py:40]: Are we sure in this case the `worker` is still valid?
Maybe we want to recreate the worker completely in this case?

File: worker/transcribee_worker/worker.py [hunk 12]
```diff
@@ -52,18 +53,13 @@ def _get_tmpfile(self, filename: str) -> Path:
             raise ValueError("`tmpdir` must be set")
         return self.tmpdir / filename
 
-    def get_document_file(self, document: ApiDocument) -> Optional[Path]:
-        logging.debug(f"Getting file. {document=}")
+    def get_document_audio(self, document: ApiDocument) -> Optional[BytesIO]:
```
Comment (rroohhh) [inline worker/transcribee_worker/worker.py:56]: Do we want to use `Optional[T]` or `T | None`?
(I am fine either way, but a consistent style is probably nice)

(Are there linters that allow custom lint rules for stuff like this?)

**Comment by rroohhh** [comment]
Apart from my comments I am fine with merging :)

**Review by rroohhh — APPROVED** [review]

File: worker/run.py (hunk 11 above)
Comment (anuejn) [reply worker/run.py:40]: I also thought about that (also validating the connection in the constructor) but I didn't see any immediate advantage and think it is fine like that for now

File: worker/transcribee_worker/worker.py (hunk 12 above)
Comment (anuejn) [reply worker/transcribee_worker/worker.py:56]: I dont know but I think our codebase does not use the | syntax anywhere yet.
I also think that it is the only python 3.10 feature we would require.

Comment (pajowu) [reply worker/transcribee_worker/worker.py:56]: I'll keep optional for now, but am open to changing it at some point if we agree to use `|` in general

File: worker/transcribee_worker/whisper_transcribe.py (hunk 7 above)
Comment (pajowu) [reply worker/transcribee_worker/whisper_transcribe.py:144]: Oops, got catch. I changed it during testing

File: worker/run.py (hunk 11 above)
Comment (pajowu) [reply worker/run.py:40]: Let's cross that bridge once we get there. Currently we also have a blanket except in `Worker.run_task` which will capture any exceptions from the actual task code, which I think would be the main source of state corruption, so this needs more work
